test/integration/consul-container/upgrade/basic_test.go

```diff
@@ -0,0 +1,324 @@
+	err, index := serviceCreate(t, client, serviceName)
+
+	ch := make(chan []*api.ServiceEntry)
+	errCh := make(chan error)
+	go func() {
+		service, q, err := client.Health().Service(serviceName, "", false, &api.QueryOptions{WaitIndex: index})
+		if q.QueryBackend != 2 {
+			err = fmt.Errorf("invalid backend for this test %d", q.QueryBackend)
+		}
+		if err != nil {
+			errCh <- err
+		} else {
+			ch <- service
+		}
+	}()
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{Name: serviceName, Port: 9998})
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	case err := <-errCh:
+		require.NoError(t, err)
+	case service := <-ch:
+		require.Len(t, service, 1)
+		require.Equal(t, serviceName, service[0].Service.Service)
+		require.Equal(t, 9998, service[0].Service.Port)
+	case <-timer.C:
+		t.Fatalf("test timeout")
+	}
+}
+
+func TestMixedServersMajorityCurrentGAClient(t *testing.T) {
+	t.Parallel()
+	var configs []consulNode.Config
+	configs = append(configs,
+		consulNode.Config{
+			HCL: `node_name="` + utils.RandName("consul-server") + `"
+					log_level="TRACE"
+					bootstrap_expect=3
+					server=true`,
+			Cmd:     []string{"agent", "-client=0.0.0.0"},
+			Version: *latestImage,
+		})
+
+	for i := 1; i < 3; i++ {
+		configs = append(configs,
+			consulNode.Config{
+				HCL: `node_name="` + utils.RandName("consul-server") + `"
+					log_level="TRACE"
+					bootstrap_expect=3
+					server=true`,
+				Cmd:     []string{"agent", "-client=0.0.0.0"},
+				Version: *curImage,
+			})
+
+	}
+
+	cluster, err := consulCluster.New(configs)
+	require.NoError(t, err)
+	defer Terminate(t, cluster)
+
+	numClients := 1
+	Clients, err := clientsCreate(numClients)
+	client := Clients[0].GetClient()
+	err = cluster.AddNodes(Clients)
+	retry.RunWith(&retry.Timer{Timeout: retryTimeout, Wait: retryFrequency}, t, func(r *retry.R) {
+		leader, err := cluster.Leader()
+		require.NoError(r, err)
+		require.NotEmpty(r, leader)
+		members, err := client.Agent().Members(false)
+		require.Len(r, members, 4)
+	})
+
+	serviceName := "api"
+	err, index := serviceCreate(t, client, serviceName)
+
+	ch := make(chan []*api.ServiceEntry)
+	errCh := make(chan error)
+	go func() {
+		service, q, err := client.Health().Service(serviceName, "", false, &api.QueryOptions{WaitIndex: index})
+		if q.QueryBackend != 2 {
+			err = fmt.Errorf("invalid backend for this test %d", q.QueryBackend)
+		}
+		if err != nil {
+			errCh <- err
+		} else {
+			ch <- service
+		}
+	}()
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{Name: serviceName, Port: 9998})
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	case err := <-errCh:
+		require.NoError(t, err)
+	case service := <-ch:
+		require.Len(t, service, 1)
+		require.Equal(t, serviceName, service[0].Service.Service)
+		require.Equal(t, 9998, service[0].Service.Port)
+	case <-timer.C:
+		t.Fatalf("test timeout")
+	}
+}
+
+func clientsCreate(numClients int) ([]consulNode.Node, error) {
```

Comment:
if we plan on putting test cases largely grouped into files like `basic_test.go`, maybe we could factor these out in some `testutils` package?

Reply:
It will very likely be the case in the future but I would wait for another test suite to be created before doing that. I don't want to extract this prematurely without having a use case for extracting it.